reference/activity_manager/pre_processing.rst

```diff
@@ -0,0 +1,85 @@
+
+We need to go deeper
+____________________
+
+The need is to apply permissions of each user on a products selection in order to acquire a new sub-selection. Then
+calculate which percentage of this sub-selection is not started to enrich, in progress or done. That is to say for each
+user and for each product in the selection apply permissions and in the user perimeter watching if there is at least one
+attribute filled or all attributes filled.
+
+Doing this synchronously is a huge task and absolutely not possible in the present state of PIM.
+
+One table to rule them all
+__________________________
+
+We found two ways to deal with this problem. First one was to calculate this detailed completeness for each user and
+cache it in a table. Benefit of this method is clear, each time we need those numbers we simply query the database: "Hey
+you, I just met Mary, what is her completeness? TODO: 42, IN_PROGRESS: 1251, DONE: 3254.". And this is crazy, result is
+instant, but the drawback, among others, is that the data is not up to date as this table is a cache. It needs to be
+calculate as often as possible but the task is so huge that with several different projects and users a night could not
+be sufficient.
+
+This first method seems not to be the chosen one. Let's recap which data we need:
+ - For each locale,
+ - For each channel,
+ - For each user,
+ - Does the product
+ - have at least an attribute filled in?
+ - or is it complete?
+
+Oh. Can you see this?
+
+Doing a table to gather all data we need to compute the completeness seems to be a good idea but we need some
+adjustments. It can't be scalable if we do this for each user.
+
+.. _format: format.html
+
+So let's remove users from this table. And now, how to retrieve the completeness from a user? In the project format_,
+we have product filters. From those filters it's easy to know which user is impacted by the products selection. To win
+some times, we pre-calculate this data linking users impacted to project at its creation. But it's not scalable too. If
+we add a user, its completeness could not be calculated because he is not linked to project yet. Problem is resolved
+linking user groups to project and not directly to users.
+
+Some data are still missing. User groups permissions are not simply about products but about attribute groups.
+
+Here are the final tables we needed to gather all data (only data that are relevant for the demonstration are shown):
+ - Pre-processing table:
+  - For each locale,
+  - For each channel,
+  - Does the attribute group,
+  - in the product,
+  - have at least an attribute filled in?
+  - or is it complete?
+
+ - Project table:
+  - locale
+  - channel
+  - product filters
+  - products (to be able to join project to pre-processing)
+  - user groups (to know who is impacted by products selection)
+
+By cross-checking pre-processing table, project table, category accesses table, attribute group accesses table we are
```

Comment:
We added an extra table for mongo DB installations which permit to know the product categories because those informations are stored in mongo. All pre processed data are stored in mysql...